gpu/gpu_info_cuda.h

```diff
@@ -6,9 +6,18 @@
 // Just enough typedef's to dlopen/dlsym for memory information
 typedef enum nvmlReturn_enum {
   NVML_SUCCESS = 0,
+  NVML_ERROR_NOT_SUPPORTED = 3,
+  NVML_ERROR_NOT_FOUND = 6,
   // Other values omitted for now...
 } nvmlReturn_t;
 typedef void *nvmlDevice_t;  // Opaque is sufficient
+
+typedef struct {
+  unsigned numDevices;
+  nvmlDevice_t **layout;
+} deviceMap_t;
```

Comment:
This is sort of solving 2 problems I believe - a cached lookup of the devices, and support for MIG virtualized subdevices.  I think it may be a bit more readable and handle the dynamic MIG device count if we created a secondary struct to represent each device.  Something along these lines maybe...

```c
typedef struct {
  nvmlDevice_t id;
  unsigned numMigs;
  nvmlDevice_t *migs; // calloc(numMigs, sizeof(nvmlDevice_t))
} deviceMap_t;
typedef struct {
  unsigned numDevices;
  deviceMap_t *devices; // calloc(numDevices, sizeof(deviceMap_t)
} deviceList_t;
```

I believe this would let us build this up dynamically as we discover the number of devices, and for each device, how many MIGs it supports.

Reply:
The host GPUs can indeed be queried for some of their capabilities, however, in k8s at least, they can not be used and other IOCTLs to them result in permission denied; for example I think I tried that at first and get-num-instances  is denied, and instead you must blindly iterate (get next) until it fails. Nor is re-allocation, or re-configuration allowed, as this is a node-level system administrator kind of thing and pods request resources using the MIG instance names only.  it would be an interesting experiment to have an H100 on a windows desktop (if you're a billionaire).